### PR TITLE
Fix to prevent data races

### PR DESCRIPTION
**Pull request by Mixaster995**
Signed-off-by: Mikhail Avramenko <avramenkomihail15@gmail.com>

## Description
Removed unnecessary fields from context to prevent data races in logs.

## Issue
fixes https://github.com/networkservicemesh/cmd-nsmgr/issues/419

### CODE REVIEW

File: internal/manager/manager.go [hunk 1]
```diff
@@ -96,7 +96,6 @@ func RunNsmgr(ctx context.Context, configuration *config.Config) error {
 
 	// Context to use for all things started in main
 	m.ctx, m.cancelFunc = context.WithCancel(ctx)
-	m.ctx = log.WithFields(m.ctx, map[string]interface{}{"cmd": "Nsmgr"})
```
Comment (denis-tingaikin) [inline internal/manager/manager.go:99]: I think this is needed for https://github.com/networkservicemesh/cmd-nsmgr/blob/main/internal/manager/manager.go#L164

Please consider using separate contexts for chain nsmgr and for logging in `RunNsmgr`

Comment (Mixaster995) [reply internal/manager/manager.go:99]: ok, done

File: internal/manager/manager.go [hunk 2]
```diff
@@ -96,8 +96,8 @@ func RunNsmgr(ctx context.Context, configuration *config.Config) error {
...
-	m.ctx = log.WithLog(m.ctx, logruslogger.New(m.ctx))
+	logCtx := log.WithLog(m.ctx, logruslogger.New(m.ctx))
+	logCtx = log.WithFields(logCtx, map[string]interface{}{"cmd": "Nsmgr"})
```
Comment (denis-tingaikin) [inline internal/manager/manager.go:100]: logger = logruslogger.New(m.ctx)

logger...

Comment (denis-tingaikin) [reply internal/manager/manager.go:100]: TODO1: add spanlogger
TODO2: use the logger in each place where we are logging in this code file.

Comment (Mixaster995) [reply internal/manager/manager.go]: ok, removed logrus, added separate context with spanlogger

**Comment by edwarnicke** [comment]
Why does trace need its on context?

**Comment by Mixaster995** [comment]
@edwarnicke you right, it is unnecessary, added some refactoring

File: main.go [hunk 3]
```diff
@@ -51,57 +48,44 @@ func main() {
...
+	traceCtx := log.WithFields(ctx, map[string]interface{}{"cmd": os.Args[:1]})
+	traceLogger, finish := utils.GetTraceLogger(traceCtx, "nsmgr")
+	defer finish()
 
 	// ********************************************************************************
 	// Debug self if necessary
 	// ********************************************************************************
 	if err := debug.Self(); err != nil {
-		log.FromContext(ctx).Infof("%s", err)
+		traceLogger.Infof("%s", err)
```
Comment (denis-tingaikin) [inline main.go:59]:
```suggestion
		logger.Infof("%s", err)
```

File: main.go [hunk 4]
```diff
@@ -51,57 +48,44 @@ func main() {
...
 	}
 
 	// ********************************************************************************
 	// Configure open tracing
 	// ********************************************************************************
 	// Enable Jaeger
 	log.EnableTracing(true)
-	jaegerCloser := jaeger.InitJaeger(ctx, "nsmgr")
+	jaegerCloser := jaeger.InitJaeger(traceCtx, "nsmgr")
```
Comment (denis-tingaikin) [inline main.go:67]:
```suggestion
	jaegerCloser := jaeger.InitJaeger(ctx, "nsmgr")
```

File: main.go [hunk 5]
```diff
@@ -51,57 +48,44 @@ func main() {
...
 	defer func() { _ = jaegerCloser.Close() }()
-	traceCtx, finish := withTraceLogger(ctx, "nsmgr")
 
 	// Get cfg from environment
 	cfg := &config.Config{}
 	if err := envconfig.Usage("nsm", cfg); err != nil {
-		logrus.Fatal(err)
+		traceLogger.Fatal(err)
```
Comment (denis-tingaikin) [inline main.go:73]:
```suggestion
		logger.Fatal(err)
```

File: main.go [hunk 6]
```diff
@@ -51,57 +48,44 @@ func main() {
...
 	}
 	if err := envconfig.Process("nsm", cfg); err != nil {
-		logrus.Fatalf("error processing cfg from env: %+v", err)
+		traceLogger.Fatalf("error processing cfg from env: %+v", err)
```
Comment (denis-tingaikin) [inline main.go:76]:
```suggestion
		logger.Fatalf("error processing cfg from env: %+v", err)
```

File: main.go [hunk 7]
```diff
@@ -51,57 +48,44 @@ func main() {
...
 	}
 
-	logrus.Infof("Using configuration: %v", cfg)
+	traceLogger.Infof("Using configuration: %v", cfg)
```
Comment (denis-tingaikin) [inline main.go:79]:
```suggestion
	logger.Infof("Using configuration: %v", cfg)
```

File: main.go [hunk 8]
```diff
@@ -51,57 +48,44 @@ func main() {
...
 
 	level, err := logrus.ParseLevel(cfg.LogLevel)
 	if err != nil {
-		logrus.Fatalf("invalid log level %s", cfg.LogLevel)
+		traceLogger.Fatalf("invalid log level %s", cfg.LogLevel)
```
Comment (denis-tingaikin) [inline main.go:83]:
```suggestion
		logger.Fatalf("invalid log level %s", cfg.LogLevel)
```

File: main.go [hunk 9]
```diff
@@ -51,57 +48,44 @@ func main() {
...
 	}
 	logrus.SetLevel(level)
 
-	// Startup is finished
-	finish()
-
-	err = manager.RunNsmgr(traceCtx, cfg)
+	err = manager.RunNsmgr(ctx, cfg)
```
Comment (denis-tingaikin) [inline main.go:87]:
```suggestion
	err = manager.RunNsmgr(ctx, logger, cfg)
```

File: internal/manager/manager.go [hunk 10]
```diff
@@ -87,20 +87,21 @@ func (m *manager) initSecurity() (err error) {
 func RunNsmgr(ctx context.Context, configuration *config.Config) error {
 	starttime := time.Now()
 
+	log.EnableTracing(true)
+	traceCtx := log.WithFields(ctx, map[string]interface{}{"cmd": "Nsmgr"})
+	traceLogger, finish := utils.GetTraceLogger(traceCtx, "start")
+	defer finish()
```
Comment (denis-tingaikin) [inline internal/manager/manager.go:93]: Let's move this to 
```suggestion
```

File: internal/manager/manager.go [hunk 11]
```diff
@@ -87,20 +87,21 @@ func (m *manager) initSecurity() (err error) {
...
+
 	m := &manager{
 		configuration: configuration,
+		traceLogger:   traceLogger,
```
Comment (denis-tingaikin) [inline internal/manager/manager.go:97]:
```suggestion
		logger:   logger,
```

File: internal/manager/manager.go [hunk 12]
```diff
@@ -87,20 +87,21 @@ func (m *manager) initSecurity() (err error) {
 func RunNsmgr(ctx context.Context, configuration *config.Config) error {
```
Comment (denis-tingaikin) [inline internal/manager/manager.go:87]:
```suggestion
func RunNsmgr(ctx context.Context, logger log.Logger, configuration *config.Config) error {
```

File: internal/manager/manager.go [hunk 13]
```diff
@@ -54,6 +53,7 @@ const (
 
 type manager struct {
 	ctx           context.Context
+	traceLogger   log.Logger
```
Comment (denis-tingaikin) [inline internal/manager/manager.go:56]:
```suggestion
	logger   log.Logger
```

File: internal/manager/manager.go [hunk 14]
```diff
@@ -54,6 +53,7 @@ const (
 
 type manager struct {
```
Comment (denis-tingaikin) [inline internal/manager/manager.go:54]: I would be suggested to remove this package. And use all this code inline. 
@edwarnicke Thoughts?

File: main.go [hunk 15]
```diff
@@ -51,57 +48,44 @@ func main() {
...
-	}
-}
-
-func withTraceLogger(ctx context.Context, operation string) (c context.Context, f func()) {
-	ctx, sLogger, span, sFinish := spanlogger.FromContext(ctx, operation)
-	ctx, lLogger, lFinish := logruslogger.FromSpan(ctx, span, operation)
-	return log.WithLog(ctx, sLogger, lLogger), func() {
-		sFinish()
-		lFinish()
+		traceLogger.Fatalf("error executing rootCmd: %v", err)
```
Comment (denis-tingaikin) [inline main.go:89]:
```suggestion
		logger.Fatalf("error executing rootCmd: %v", err)
```

**Review by denis-tingaikin — CHANGES_REQUESTED** [review]

File: main.go [hunk 16]
```diff
@@ -51,57 +48,44 @@ func main() {
 	)
 	defer cancel()
 
-	logrus.SetFormatter(&nested.Formatter{})
-	ctx = log.WithFields(ctx, map[string]interface{}{"cmd": os.Args[:1]})
-	ctx = log.WithLog(ctx, logruslogger.New(ctx))
+	traceCtx := log.WithFields(ctx, map[string]interface{}{"cmd": os.Args[:1]})
+	traceLogger, finish := utils.GetTraceLogger(traceCtx, "nsmgr")
```
[the remark this answers is not in the document]
Comment (denis-tingaikin) [reply main.go:52]: Note: I think we need to expose group func from sdk

[the remark this answers is not in the document]
Comment (Mixaster995) [reply main.go:52]: should i prepare separate PR to sdk for this?

[the remark this answers is not in the document]
Comment (denis-tingaikin) [reply main.go:52]: @Mixaster995 yes

File: internal/manager/manager.go (hunk 14 above)
Comment (edwarnicke) [reply internal/manager/manager.go:54]: In the generic I agree... not sure whether its necessary before the release or not though.  How much work would it be and how much risk would it entail?

**Comment by edwarnicke** [comment]
I'm a bit confused how this connects to concurrent map panics in #419, could you explain?

**Comment by denis-tingaikin** [comment]
@edwarnicke 

Sure. We are setting a setlogoptions at the top of the application. 

So it being conflicts with each connection here
https://github.com/networkservicemesh/sdk/blob/main/pkg/networkservice/common/setlogoption/server.go#L47

Imagine a few goroutines that are trying to work with this map.

**Comment by denis-tingaikin** [comment]
> Why does our logging involve a map with this risk? Does vanilla logrus have these issues?

No,  vanilla logrus doesn't have this issue. Currently, we are incorrect in using our logger from sdk.

**Comment by edwarnicke** [comment]
Got it.  I look forward to seeing this land :)

File: internal/manager/manager.go (hunk 14 above)
Comment (denis-tingaikin) [reply internal/manager/manager.go:54]: I think you're correct. Let's consider this after release.

**Comment by Mixaster995** [comment]
@denis-tingaikin made requested changes - please, take a look

File: main.go [hunk 17]
```diff
@@ -51,57 +49,46 @@ func main() {
...
 
 	level, err := logrus.ParseLevel(cfg.LogLevel)
 	if err != nil {
-		logrus.Fatalf("invalid log level %s", cfg.LogLevel)
+		logger.Fatalf("invalid log level %s", cfg.LogLevel)
 	}
 	logrus.SetLevel(level)
 
-	// Startup is finished
-	finish()
```
Comment (denis-tingaikin) [inline main.go:92]: Hmm, this can be a major. Can we run here `sFinish()`?

Comment (denis-tingaikin) [reply main.go:92]: Yes, we can re-recreate the span logger for the next operation.

Comment (Mixaster995) [reply main.go:92]: ok, done

**Review by denis-tingaikin — APPROVED** [review]

**Pull request merged**